src/field/HelixStepper.hh

```diff
@@ -0,0 +1,164 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file HelixStepper.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Types.hh"
+
+#include "Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * The Helix Stepper
+ *
+ * The analytical solution for the motion of a charged particle in a uniform
+ * magnetic field.
+ */
+template<class EquationT>
+class HelixStepper
+{
+  public:
+    //!@{
+    //! Type aliases
+    using Result = StepperResult;
+    //!@}
+
+  public:
+    // Construct with the equation of motion
+    CELER_FUNCTION
+    HelixStepper(const EquationT& eq) : equation_(eq) {}
+
+    // Adaptive step size control
+    CELER_FUNCTION auto operator()(real_type step, const OdeState& beg_state)
+        -> Result;
+
+  private:
+    //// DATA ////
+
+    // Equation of the motion
+    const EquationT& equation_;
+
+    //// HELPER FUNCTIONS ////
+
+    // Analytical solution for a given step along a helix trajectory
+    CELER_FUNCTION OdeState move(real_type       step,
+                                 real_type       radius,
+                                 real_type       helicity,
+                                 const OdeState& beg_state,
+                                 const OdeState& rhs);
+
+    //// COMMON PROPERTIES ////
+
+    static CELER_CONSTEXPR_FUNCTION real_type tolerance() { return 1e-10; }
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * An explicit helix stepper with analytical solutions at the end and the
+ * middle point for a given step.
+ *
+ * Assuming that the magnetic field is uniform and chosen to be parallel to
+ * the z-axis, \f$B = (0, 0, B_z)\f$, without loss of generality, the motion of
+ * a charged particle is described by a helix trajectory. For this algorithm,
+ * the radius of the helix, \f$R = m gamma v/(qB)\f$ and the helicity, defined
+ * as \f$ -sign(q B_z)\f$ are evaluated through the right hand side of the ODE
+ * equation where q is the charge of the particle.
+ */
+template<class E>
+CELER_FUNCTION auto
+HelixStepper<E>::operator()(real_type step, const OdeState& beg_state)
+    -> Result
+{
+    Result result;
+
+    // Evaluate the right hand side of the equation
+    OdeState rhs = equation_(beg_state);
+
+    // Calculate the radius of the helix
+    real_type radius
+        = sqrt(ipow<2>(norm(beg_state.mom)) - ipow<2>(beg_state.mom[2]))
+          / norm(rhs.mom);
+
+    // Set the helicity: 1(-1) for negative(positive) charge with Bz > 0
+    real_type helicity = (rhs.mom[0] / rhs.pos[1] < 0) ? 1 : -1;
```

Comment:
Okay, I added as suggested.  Isn't the conversion from int to real_type expensive too?

Reply:
Well, looks like you're right! I ran a little test to see what the compiler did in both cases, and having the conditional here is definitely less expensive (in register usage and instruction count) than `2 * h - 1`. However, one final test case suggests the best choice would be keeping the `Helicity` bool and using `real_type y = (Helicity == positive ? x : -x);` is the best case of all. I think in your updated code it's better to pass in Helicity as part of the interface to `move`, rather than converting it to an `int` first.

https://github.com/celeritas-project/cuda-test-snippets/tree/master/signflip

https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#comparison-and-selection-instructions-selp

Reply:
Nice test/study (a good practice for a tiny performance improvement! - in Korean proverb says, "Many a tiny makes a big mountain").  I dropped `is_sign` function and modified the use of Helicity bool as suggested.